MechJeb2/LandingAutopilot/FinalDescent.cs

```diff
@@ -81,26 +83,42 @@ public override AutopilotStep Drive(FlightCtrlState s)
                         double terrainRadius = mainBody.Radius + mainBody.TerrainAltitude(estimatedLandingPosition);
                         aggressivePolicy = new GravityTurnDescentSpeedPolicy(terrainRadius, mainBody.GeeASL * 9.81, vesselState.limitedMaxThrustAccel);  // this constant policy creation is wastefull...
                         core.thrust.trans_spd_act = (float)(aggressivePolicy.MaxAllowedSpeed(vesselState.CoM - mainBody.position, vesselState.surfaceVelocity));
+                        thrustAt200Meters = (float)vesselState.limitedMaxThrustAccel; // this gets updated until we are below 200 meters
                     }
                 }
                 else
                 {
-                    // last 200 meters:
-                    core.thrust.trans_spd_act = -Mathf.Lerp(0, (float)Math.Sqrt((vesselState.limitedMaxThrustAccel - vesselState.localg) * 2 * 200) * 0.90F, (float)minalt / 200);
+                    float previous_trans_spd_act = Math.Abs(core.thrust.trans_spd_act); // avoid issues going from KEEP_SURFACE mode to KEEP_VERTICAL mode
 
-                    // take into account desired landing speed:
-                    core.thrust.trans_spd_act = (float)Math.Min(-core.landing.touchdownSpeed, core.thrust.trans_spd_act);
+                    // Last 200 meters, at this point the rocket has a TWR that will rise rapidly, so be sure to use the same policy based on an older
+                    // TWR. Otherwise we would suddenly get a large jump in desired speed leading to a landing that is too fast.
+                    core.thrust.trans_spd_act = Mathf.Lerp(0, (float)Math.Sqrt((thrustAt200Meters - vesselState.localg) * 2 * 200) * 0.90F, (float)minalt / 200);
+
+                    // Sometimes during the descent we end up going up a bit due to overthrusting during breaking, avoid thrusting up even more and destabilizing the rocket
+                    core.thrust.trans_spd_act = (float)Math.Min(previous_trans_spd_act, core.thrust.trans_spd_act);
 
-//                    core.thrust.tmode = MechJebModuleThrustController.TMode.KEEP_VERTICAL;
-//                    core.thrust.trans_kill_h = true;
+                    // take into account desired landing speed:
+                    core.thrust.trans_spd_act = (float)Math.Max(core.landing.touchdownSpeed, core.thrust.trans_spd_act);
 
-//                    if (Math.Abs(Vector3d.Angle(-vessel.surfaceVelocity, vesselState.up)) < 10)
+                    // Prevent that we switch back from Vertical mode to KeepSurface mode
+                    // When that happens the rocket will start tilting and end up falling over
                     if (vesselState.speedSurfaceHorizontal < 5)
+                    {
+                        forceVerticalMode = true;
+                    }
```

Comment:
this is probably a bit flawed, the algorithm should really drive horizontal velocity to zero with main engines at 200m or so, with some clamps to make it not do silly spins hunting for a perfect zero speed -- then it can drop straight down once the horizontal speed is zero.